pkg/apis/core/types.go

```diff
@@ -4767,3 +4773,47 @@ const (
 	// DefaultHardPodAffinityWeight defines the weight of the implicit PreferredDuringScheduling affinity rule.
 	DefaultHardPodAffinitySymmetricWeight int32 = 1
 )
+
+type UnsatisfiableConstraintResponse string
+
+const (
+	// DoNotSchedule instructs the scheduler not to schedule the pod
+	// when constraints are not satisfied.
+	DoNotSchedule UnsatisfiableConstraintResponse = "DoNotSchedule"
+	// ScheduleAnyway instructs the scheduler to schedule the pod
+	// even if constraints are not satisfied.
+	ScheduleAnyway UnsatisfiableConstraintResponse = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
+	// It's the max difference between the number of matching pods in any two
+	// topology domains against a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// spread as 1/1/0:
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required value. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
+	// TopologyKey is the key of node labels. Nodes that have a label with this key
+	// and the value of the label is identical are considered in the same topology.
+	// We consider each <key, value> as a "bucket", and try to put balanced number
+	// of pods into each bucket.
+	// It's a required field.
+	TopologyKey string
+	// WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+	// the spread constraint.
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintResponse
+	// LabelSelector is used to find matching pods.
+	// Pods that match this label selector are counted to determine the number of pods
+	// in their corresponding topology domain.
+	// +optional
+	LabelSelector *metav1.LabelSelector
```

Comment:
i am trying to remember what was the actual use case for the label selector ?

Reply:
Another way of asking this is why would this label selector be different then the deployments own label selector ? I see two scenarios:-
1: i want to spread all pods of a workload controller (deployment, RS, statefulset). In this case we dont need a selector
2: i want to spread all pods matching a label. This is more common in CRD controllers which deploy a pod and ditch the regular core controllers.
I feel like we need a model where specifying this labelSelector for the first case should not be mandatory and should be derivable from the uber object . @Huang-Wei @bsalamat  wdyt ?

Reply:
A reasonable model is to not impose limitations in the beginning unless we have a strong reason (e.g. perf) to do that.

Reply:
@Huang-Wei  we discussed this in one of our sig-scheduling meetings to deduce the labelselector from its owning Deployment if one is omitted from the topologySpreadConstraint . This would make it painless for majority of the users who just want to spread the pods of their Deployment /Statefulset and wont have to specify the LabelSelector. I can create a issue on this if you want or if you want to address as part of the other implementation PR's you have.